Core/UserAgentManager.swift

```diff
@@ -124,24 +145,160 @@ struct UserAgent {
+
+    private func defaultPolicy(forConfig config: PrivacyConfiguration) -> DefaultPolicy {
+        let uaSettings = config.settings(for: .customUserAgent)
+        guard let policy = uaSettings[Constants.defaultPolicyConfigKey] as? String else { return .ddg }
+
+        return DefaultPolicy(rawValue: policy) ?? .ddg
+    }
+
+    private func ddgDefaultSites(forConfig config: PrivacyConfiguration) -> [String] {
+        let uaSettings = config.settings(for: .customUserAgent)
+        let defaultSitesObjs = uaSettings[Constants.ddgDefaultSitesConfigKey] as? [[String: String]] ?? []
+
+        return defaultSitesObjs.map { $0[Constants.uaOmitDomainConfigKey] ?? "" }
+    }
+
+    private func ddgFixedSites(forConfig config: PrivacyConfiguration) -> [String] {
+        let uaSettings = config.settings(for: .customUserAgent)
+        let fixedSitesObjs = uaSettings[Constants.ddgFixedSitesConfigKey] as? [[String: String]] ?? []
+
+        return fixedSitesObjs.map { $0[Constants.uaOmitDomainConfigKey] ?? "" }
+    }
+
+    private func closestUserAgentVersions(forConfig config: PrivacyConfiguration) -> [String] {
+        let uaSettings = config.settings(for: .customUserAgent)
+        let closestUserAgent = uaSettings[Constants.closestUserAgentConfigKey] as? [String: Any] ?? [:]
+        let versions = closestUserAgent[Constants.uaVersionsKey] as? [String] ?? []
+        return versions
+    }
+
+    private func ddgFixedUserAgentVersions(forConfig config: PrivacyConfiguration) -> [String] {
+        let uaSettings = config.settings(for: .customUserAgent)
+        let fixedUserAgent = uaSettings[Constants.ddgFixedUserAgentConfigKey] as? [String: Any] ?? [:]
+        let versions = fixedUserAgent[Constants.uaVersionsKey] as? [String] ?? []
+        return versions
+    }
+
+    public func agent(forUrl url: URL?,
+                      isDesktop: Bool,
                       privacyConfig: PrivacyConfiguration = ContentBlocking.shared.privacyConfigurationManager.privacyConfig) -> String {
+
+        guard privacyConfig.isEnabled(featureKey: .customUserAgent) else { return oldLogic(forUrl: url,
+                                                                                           isDesktop: isDesktop,
+                                                                                           privacyConfig: privacyConfig) }
+
+        if ddgDefaultSites(forConfig: privacyConfig).contains(where: { domain in
+            url?.isPart(ofDomain: domain) ?? false
+        }) { return oldLogic(forUrl: url, isDesktop: isDesktop, privacyConfig: privacyConfig) }
+
+        if ddgFixedSites(forConfig: privacyConfig).contains(where: { domain in
+            url?.isPart(ofDomain: domain) ?? false
+        }) { return ddgFixedLogic(forUrl: url, isDesktop: isDesktop, privacyConfig: privacyConfig) }
+
+        if closestUserAgentVersions(forConfig: privacyConfig).contains(statistics.atbWeek ?? "") {
+            if canUseClosestLogic {
```

Comment:
You could bury this check within the `closestLogic` as you have for `ddgFixedLogic` and just keep the single other check within `ddgFixedLogic` to keep the fall-back behaviour correct.

But also... we're going to rip this out quickly again hopefully too; so feel free to ignore.

Reply:
That was my initial idea, but it caused a side effect that made one test fail - when the closest user agent "could not be used" because of old version it fallbacked to oldUserAgentLogic which returned the old UA string which contained "DuckDuckGo/7" component. So inside ddgFixedLogic we were appending yet another DuckDuckGo/7 component.

So my idea was to make it as specific as it can be, because we want to remove it soon.

Reply:
Yeah you'd still need the check in `ddgFixedLogic` to keep the same structure I think that should work (but also not tried).